### ISSUE TITLE
Execute program with file name(s) as argument

### ISSUE THREAD

**Issue by oyarsa** [Canop/broot#695]
This seems like an obvious thing, so I apologise if it's duplicated somewhere.

We can define verbs that execute arbitrary commands, but there doesn't seem to be a way to run commands from within broot itself. I don't want to define every single command I might use as a verb; it should be possible to run something like `!vim %` to run Vim with the current file being highlighted passed as an argument.

This should also work with the staging area: if multiple files are selected, they should expand into the arguments. E.g., if the staging area contains files `a.txt` and `b.txt`, `!vim %` would execute `vim a.txt b.txt`.

**Comment by Canop** [comment]
Today, it's possible to define a verb like this:

```Hjson
    {
        invocation: "run {exec}"
        shortcut: r
        execution: "{exec} {file}"
        leave_broot: false
    }
```

With such a verb you can type `:r vim` and it launches vim on the selected file.

Allowing `:!vim` is probably doable.

**Comment by oyarsa** [comment]
This works great with single files, thank you. However, this doesn't behave as expected with the staging area. If I stage a few files and run `r vim`, I get individual instances of vim for each file instead of one vim instance with all files. I think instead of doing `vim file1 file2`, it's doing `vim file1; vim file2`.

**Comment by Canop** [comment]
Yes, the behavior of verbs on the staging area today is to execute sequentially.
I'm pondering the best way to manage that. Either a {files} argument for the execution or a verb parameter describing the behavior to use for multiple selections.

**Comment by thealio** [comment]
> Today, it's possible to define a verb like this:
> 
> ```
>     {
>         invocation: "run {exec}"
>         shortcut: r
>         execution: "{exec} {file}"
>         leave_broot: false
>     }
> ```
> 
> With such a verb you can type `:r vim` and it launches vim on the selected file.
> 
> Allowing `:!vim` is probably doable.

Hi,
I'm trying to use this for launch my scripts as following
```

 # This verb lets you run an executable with ctrl-r
   {
        invocation: "run {exec}"
        shortcut: r
        key: ctrl-r
        execution: "{exec} {file}"
        leave_broot: false
    }
```

and , on broot , I select the script , then I write
`
:r exec`

But the command is not successful . On bash `exec myscript.sh` works
I'd like to be able to do the equivalent of `./mysript.sh` on the shell

**Comment by Canop** [comment]
It's a little more complex for a shell script than for an executable. What's your shell script doing ? Can you post it here (or a shortened variant) ?

**Comment by thealio** [comment]
> It's a little more complex for a shell script than for an executable. What's your shell script doing ? Can you post it here (or a shortened variant) ?

Every bash script I run don't work
Even a simple test script
```

#!/bin/bash

echo "This is a test"
sleep 1
echo "It should work well"
```

**Comment by Canop** [comment]
You have to call an executable, to run your script. For example:

    {
        invocation: hello
        external: "sh -e /path/to/hello.sh"
        leave_broot: false
    }

**Comment by Canop** [comment]
As I realize this may not be obvious, I'm adding a section in the doc: https://dystroy.org/broot/conf_verbs/#call-shell-scripts

**Comment by thealio** [comment]
> You have to call an executable, to run your script. For example:
> 
> ```
> {
>     invocation: hello
>     external: "sh -e /path/to/hello.sh"
>     leave_broot: false
> }
> ```
> 
> If your script takes arguments:
> 
> ```
> {
>     invocation: hello
>     external: "sh -e /path/to/do_something.sh {file}"
>     leave_broot: false
> }
> ```

Hi, 
thank you for your efforts about my request.
Following your guide, I was able to archive scripts launching as following

  ```
  # This verb lets you run an executable with ctrl-r
  # (on exit you'll be back in broot)
    {
        invocation: "run"
        shortcut: ru
        key: ctrl-r
        external: "sh -e {file}"
        leave_broot: false
    }
```
 I have another question, for making script launching even better.
Is possible to launch scripts with their flags?
For example , supposing the script `hello.sh` can be launched with flag `-h`  for help as `hello.sh -h` .
Piratically  a way to add " `-h`" or every other word at the end of the script name, typing it in broot

**Comment by Canop** [comment]
You mean type the flag in broot ? Yes:
```

  {
      invocation: "r"
      external: ["sh" "-e" "{file}"]
      leave_broot: false
  }
  {
      invocation: "rf {flag}"
      external: ["sh" "-e" "{file}" "-{flag}"]
      leave_broot: false
  }
```

Select your executable then type 'r' and enter to execute it without flag, or 'rf h' to execute it with '-h'.

**Comment by thealio** [comment]
> invocation: "r"
>       external:

Hi 
it works perfectly! Really nice!

my final code is the following

 ```
   # This verb lets you run an executable shell script with ctrl-r
    # (on exit you'll be back in broot)
    {
        invocation: "run"
        shortcut: ru
        key: ctrl-r
        external: "sh -e {file}"
        leave_broot: false
    }
 
    # Run shell script with flag --> rf [flag]
    {
        invocation: "runf {flag}"
        shortcut: rf
        external: "sh -e {file} -{flag}"
        leave_broot: false
    }
    

```
This issue can be consider as resolved. 
Many thanks !

**Comment by Martin1887** [comment]
This is very interesting and it has a lot of potential applications. I think it should be highlighted in the web or another place (maybe another blog post?).

In my opinion, it would be more flexible if you could type any command and use placeholders to indicate file and (parent if it is a file) directory. That would be a great addition to broot!

In a future stage, commands could be suggested (à la vim commands) in relation to the MIME type of the file, this would be related to #823.

Thanks.

**Comment by Canop** [comment]
> and use placeholders to indicate file and (parent if it is a file) directory. That would be a great addition to broot!

I'm sorry but I don't follow you here. Could you give some examples ?

**Comment by Martin1887** [comment]
Of course. It could be like the syntax for creating verbs in hjson/TOML files but directly into broot for one-time commands that you will probably not execute anymore (so configuring the verb does not pay off). Some examples:

```
 tar -xzvf {file}
 mv {file} {file-name}.old
 okular {file} (supposing your default PDF reader is evince for instance)
 codium {directory}
```

We could name them something like 'live verbs'.

For the commands recommendation, it could be triggered when typing and optionally using and special shortcut (probably `Tab`), and it could be based on the MIME type of the selected file for a better user experience. But this is only a nice addition that could be added in the future.

Thanks.

**Comment by Canop** [comment]
A generic verb does this:

    {
        invocation: "run {cmd}"
        shell_command: "{cmd} {file:space-separated}"
        leave_broot: false
    }

Then `:run chmod +x` or `:run tar czf out.tgz` applies to the selection, staged files included, and pipes work since it goes through a shell (1.59). Placeholders inside what you type (`okular {file}`) aren't expanded, only those of the verb definition are. Closing.
